File: python/rules/domain-modeling/max-parameters.py
from typing import List
from typing_extensions import override

# ok: python-max-parameters-warning
# ok: python-max-parameters-error
def calculate_sum(a: int, b: int) -> int:
    return a + b

# ruleid: python-max-parameters-warning
def build_point(x: float, y: float, z: float) -> tuple:
    return (x, y, z)

# ruleid: python-max-parameters-error
def create_user(name: str, email: str, age: int, role: str) -> dict:
    return {"name": name, "email": email, "age": age, "role": role}

class BaseCustomerController:
    pass

class CustomerController(BaseCustomerController):
    # ok: python-max-parameters-warning
    # ok: python-max-parameters-error
    def __init__(self, a: str, b: str, c: str, d: str, e: str):
        self.a = a
        self.b = b
        self.c = c
        self.d = d
        self.e = e

    # ok: python-max-parameters-warning
    # ok: python-max-parameters-error
    def find_user(self, user_id: str, tenant_id: str) -> str:
        return user_id

    # ruleid: python-max-parameters-warning
    def update_user(self, user_id: str, first_name: str, last_name: str) -> str:
        return user_id

    # ruleid: python-max-parameters-error
    def register_user(self, username: str, email: str, address: str, phone: str) -> str:
        return username

    # ok: python-max-parameters-warning
    # ok: python-max-parameters-error
    @override
    async def search_customers(
        self,
        dot_number: str,
        name: str,
        street: str,
        zip: str,
        city: str,
        state: str,
        country: str,
    ) -> List[str]:
        return []

    # ok: python-max-parameters-warning
    # ok: python-max-parameters-error
    @override
    def filter_triadic(self, a: str, b: str, c: str) -> str:
        return a

File: csharp/rules/domain-modeling/max-parameters.cs
namespace Thruput.Semgrep.Tests
{
    public class ParameterCountTests
    {
        // ok: csharp-max-parameters-warning
        // ok: csharp-max-parameters-error
        public ParameterCountTests(string a, string b, string c, string d, string e)
        {
        }

        // ok: csharp-max-parameters-warning
        // ok: csharp-max-parameters-error
        public int Add(int a, int b)
        {
            return a + b;
        }

        // ruleid: csharp-max-parameters-warning
        public void BuildPoint(int x, int y, int z)
        {
        }

        // ruleid: csharp-max-parameters-warning
        public int ExpressionThree(int a, int b, int c) => a + b + c;

        // ruleid: csharp-max-parameters-error
        public void CreateUser(string name, string email, int age, string role)
        {
        }

        // ruleid: csharp-max-parameters-error
        public void SendNotification(string user, string title, string body, string channel, bool priority)
        {
        }

        // ok: csharp-max-parameters-warning
        // ok: csharp-max-parameters-error
        [HttpGet("search")]
        public object SearchCustomers(string name, string street, string city, string state, string country)
        {
            return new object();
        }
    }

    public class HttpGetAttribute : System.Attribute
    {
        public HttpGetAttribute(string template) { }
    }
}
